### ISSUE TITLE
jQuery Fundamentals

### ISSUE THREAD

**Issue by neight-allen** [turingschool/intermission-assignments#87]
From Bocoup's [jQuery Fundamentals](http://jqfundamentals.com).

Respond in the comments of this issue with:
1. What is something you learned that was particularly surprising/interesting?
2. What was something you already knew?
3. Was there anything you feel you still don't understand?

Feel free to ask any questions about jQuery fundamentals here as well.

**Comment by kjs222** [comment]
1. the serialize method to use on ajax post requests with forms was new to me.  
2. I knew most of this content to some degree, but definitely learned a few new things here and there. 
3. I alternate between using this and event.target a lot with event handling and I need to understand the distinction between them better.  I don't think this resource went into that topic at all, or in any depth.  I also need to re-review event bubbling.

**Comment by ryanflach** [comment]
1. Function expressions are preferred over function declaration due to hoisting. Named function expressions can make debugging somewhat easier, but they do not work correctly in IE < 9, and may introduce bugs of their own. Plain function expressions are safest.
2. Outside of some basics (e.g., variable names that begin with `$` typically refer to jQuery objects; AJAX `get` and `post` convenience methods), this was all good reinforcement or new material.
3. An additional dive into event bubbling and event delegation would be beneficial. I'm realizing that I may have utilized both in my personal project without understanding that they were taking place or any potential consequences.

Additional questions:
- Cross-origin resource sharing: what is it, and what is its practical application?
- Promises: would love to know more about these and get clarification on exactly what they are (it seems as though they are the return value?).
- Deferreds: after the section in the AJAX chapter, I'm still somewhat unclear on these, as this reading was my first encounter with them.

**Comment by dtinianow** [comment]
1. The most interesting section to me was the one about AJAX requests.  I thought it was especially interesting how many different ways there are to do similar things, for example using jqXHR vs deferreds,  using a convenience method vs AJAX method, passing in an object vs passing in a URL and optional object to the AJAX request, etc. 
2. I knew most of the basic Javascript syntax and have already seen many of the jQuery methods.  Much of the AJAX section was new to me, as was the section on event bubbling and delegation.
3. I didn't get a chance to use AJAX in my personal project, so my understanding of it is a whole is pretty hazy.  I know I need to spend some time actually using it to better understand it.  I was especially confused about the section on using deferreds and it's related methods.  I could also use clarification on event bubbling and delegation.

**Comment by kbs5280** [comment]
Respond in the comments of this issue with:
1. What is something you learned that was particularly surprising/interesting?
   The section on AJAX is concise and explains well the process of calls and potential issues. _X is for JSON_ and _JSONP_ are both interesting sections. I'm particularly interested in getting to work with events and effects.
2. What was something you already knew?
   I already understood the first section on JavaScript basics, but the rest of the issue offered a majority of new information or explanations. 
3. Was there anything you feel you still don't understand?
   I feel comfortable with most of the concepts in theory, but in every area there is a lot to learn and put together in practice. The idea of traversal isn't conceptually difficult but working it into an application and making it functional and practical seems challenging. Like many programming concepts, jQuery seems like a skill that will be more accessible and understandable with practice and implementation.

**Comment by Automatic365** [comment]
1. The material on AJAX was all pretty cool to me, and all of the event delegation. 
2. The Javascript basics were all familiar to me. 
3. All of it seems straightforward in theory, but it will be fuzzy to me until I actually play around with it.

**Comment by allindow** [comment]
1. What is something you learned that was particularly surprising/interesting?
   **declaring a variable without var makes it a global variable even within a function. Creating html elements with jquery too - I feel like I already knew that one but maybe forgot?**
2. What was something you already knew? 
   **The beginning of the jQuery section about selectors**
3. Was there anything you feel you still don't understand?
   **I have a hard time grasping concepts just reading about them, so while a lot of it sounded familiar, a lot of it probably won't solidify until I've actually practiced with it.**

**Comment by zackforbing** [comment]
**1. What is something you learned that was particularly surprising/interesting?** global variables, using `.length` as an existence qualifier for jQuery objects since `0` is falsy, implicit iteration over jQuery objects
**2. What was something you already knew?** a lot of the JS and JQ stuff, due to previous Turing lessons and tutorials (such as `$(document).ready()`, using css selectors in JQ, JS naming conventions, JS structure basics, etc), calling `$.ajax`
**3. Was there anything you feel you still don't understand?** I definitely will feel more confident about these concepts when we put them into practice.

**Comment by chompasina** [comment]
1. What is something you learned that was particularly surprising/interesting? 
   I hadn't thought about 'this' changing context and therefore the object it refers to changing as well when the object methods are stored in a variable.  I was also surprised by the difference between .is and .not: .is() returns a boolean, while .not() returns a new jQuery object. I'm intrigued by event bubbling and event delegation but will have to study the concepts more before sharing my reflections.
2. What was something you already knew?
   I knew most of the jQuery selectors and basics of Javascript.
3. Was there anything you feel you still don't understand?
   As I mentioned, event bubbling and delegation are unclear to me as of yet.  I'm also not sure I could traversing the DOM or use methods like .innerHTML() without referencing these materials. The latter half of the AJAX section felt very unclear to me, particularly deferreds. In general, I wish I had been able to practice with these concepts. The sandbox doesn't seem very illuminating for most of the jQuery part, but perhaps I was doing something wrong. I'm looking forward to playing around with the concepts with real code.

**Comment by MsJennyGiraffe** [comment]
1. What is something you learned that was particularly surprising/interesting?
   I think it's really interesting learning more about the multitude of different ways there are to manipulate and traverse through html using jquery, getting a better idea of what it can all do, and getting more clarification on things we have used in class.
2. What was something you already knew?
   I already knew most of the basics of javascript, and how to very basically select and alter elements using jquery.
3. Was there anything you feel you still don't understand?
   No, I feel it was well explained and will be a great resource when it becomes project time.

**Comment by deborahleehamel** [comment]
## What is something you learned that was particularly surprising/interesting?

Interesting how 'hoisting' is the JavaScript interpreters action of moving all variable and function declarationss to the top of the current scope. Think I will just assign functions to variables to avoid these potential bugs if I can.

you can force the meaning of 'this' to be what you want it to be using .call() or .apply() method on the function itself

## What was something you already knew?

basic concepts of Javascript; variables, functions, objects, object methods, arrays

## Was there anything you feel you still don't understand?

object prototypes; constructor functions and deferreds

Like to understand getting single elements from a selection better with '.innerHTML' & .eq();
traversing and manipulating the DOM(want to get more hands on practice); preventing default behavior and event bubblilng

AJAX - reading concepts is one thing, knowing how to apply and execute contextually in the most appropriate ways, yet another challenge

**Comment by roscalabrin** [comment]
**1. What is something you learned that was particularly surprising/interesting?**
That you can force the meaning of ‘this’ using the.call() or .apply() methods.
That $.ajax method returns before the request is finished (meaning before the success callback runs). 

 **2. What was something you already knew?**
- Most of the getters and setters methods
- The ability to chain methods 
- That JSON is a string representation of an object 

 **3. Was there anything you feel you still don't understand?**
Creating your own deferreds.

**Comment by karinamzalez** [comment]
- What is something you learned that was particularly surprising/interesting?
  I thought it was interesting that math in JavaScript wont be particularly accurate... I wonder what is happening in JavaScript internally to make that happen... how does being 'a loosely typed language' affect your mathematical accuracy when it's the same machine processing the language?  I also thought it was interesting how, because ajax is asynchronous, that can cause broken code if not careful. 
- What was something you already knew?
  I was already familiar with the basics of JavaScript and JQuery selectors in general but.. I already knew how to write a ternary operator and looking at the JavaScript version of a one-line conditional was neat-- it's so similar to the one in Ruby! It feels as if I'm about to embark upon a journey of learning a variation of something I have more-or-less already learned and that is reassuring. 
-  Was there anything you feel you still don't understand?
  Deferreds.

**Comment by cconcannon** [comment]
I didn't realize that when assigning a function call to a variable, "this" no longer refers to the function itself. I also didn't realize that it is best practice in JavaScript (because of strange handling of math) to convert decimals to integers before operations, then convert back to decimals afterward. I also did not know that jQuery events can be namespaced. 

I already knew about the assignment of variables within functions and scope of variables to functions. I also knew about jQuery selectors.

I still don't quite understand how to deal with the asynchronous nature of jQuery - deferred's and pipes

**Comment by 5Notes** [comment]
1. What is something you learned that was particularly surprising/interesting?
   jQuery has a 'detach' method.

 2. What was something you already knew?
jQuery has the ability to chain methods together.
1. Was there anything you feel you still don't understand?
   The exact uses of the 'pipe' method for AJAX.

**Comment by matthewrpacker** [comment]
1.  It was interesting how the `.serialize()` method could take form input and convert it into query string format.  This seems like it could really help on future projects.
2.  You can access an object's properties using either dot notation or bracket notation.
3.  The `.detach()` method to temporarily remove elements from the page.  I'm not clear on when this would be appropriate to use.

**Comment by Laner12** [comment]
1. What is something you learned that was particularly surprising/interesting?
   This can have different values. One for Javascript and one for JQuery
2. What was something you already knew?
   Basic Javascript functions and basic JQuery selectors
3. Was there anything you feel you still don't understand?
   Not as a whole but actually using these concepts would help me solidify them.

**Comment by twhitinger** [comment]
1. What is something you learned that was particularly surprising/interesting?
   Using the blur event to trigger an action when an element loses focus, used it to save editable    forms
2. What was something you already knew?
   javascript basics
3. Was there anything you feel you still don't understand?
   application for all of the jQuery methods to make my life easier.
